source/server/drain_manager_impl.cc

```diff
@@ -52,15 +80,74 @@ bool DrainManagerImpl::drainClose() const {
          (server_.api().randomGenerator().random() % server_.options().drainTime().count());
 }
 
+Common::CallbackHandlePtr DrainManagerImpl::addOnDrainCloseCb(DrainCloseCb cb) const {
+  ASSERT(dispatcher_.isThreadSafe());
+
+  if (draining_) {
+    const MonotonicTime current_time = dispatcher_.timeSource().monotonicTime();
+
+    // If we've already passed our deadline, or we have an immediate drain-strategy, call the
+    // callback with no delay specified
+    if (current_time > drain_deadline_ ||
+        server_.options().drainStrategy() == Server::DrainStrategy::Immediate) {
```

Comment:
Consider factoring this logic into the computation of remaining_time.  Something like:

```
std::chrono::milliseconds remaining_time =
  (server_.options().drainStrategy() != Server::DrainStrategy::Immediate && current_time < drain_deadline_) ?
  std::chrono::duration_cast<std::chrono::milliseconds>(drain_deadline_ - current_time)  :
  std::chrono::milliseconds{0};
cb(remaining_time);
return nullptr;
```

Reply:
done!